mava/utils/checkpointing.py

```diff
@@ -0,0 +1,200 @@
+        timestamp_str = (
+            timestamp_override if timestamp_override else datetime.now().strftime("%Y%m%d%H%M%S")
+        )
+
+        options = orbax.checkpoint.CheckpointManagerOptions(
+            create=True,
+            best_fn=lambda x: x["episode_return"],
+            best_mode="max",
+            save_interval_steps=save_interval_steps,
+            max_to_keep=max_to_keep,
+            keep_period=keep_period,
+        )
+
+        def get_json_ready(obj: Any) -> Any:
+            if not isinstance(obj, (bool, str, int, float, type(None))):
+                return str(obj)
+            else:
+                return obj
+
+        metadata_json_ready = tree_map(get_json_ready, metadata)
+
+        self._manager = orbax.checkpoint.CheckpointManager(
+            directory=os.path.join(os.getcwd(), rel_dir, model_name, timestamp_str),
+            checkpointers=orbax_checkpointer,
+            options=options,
+            metadata={
+                "checkpointer_version": CHECKPOINTER_VERSION,
+                **(metadata_json_ready if metadata_json_ready is not None else {}),
+            },
+        )
+
+        # Don't log checkpointing messages (at INFO level)
+        absl_logging.set_verbosity(absl_logging.WARNING)
+
+    def save(
+        self,
+        timestep: int,
+        unreplicated_learner_state: Union[LearnerState, RNNLearnerState],
+        episode_return: Numeric = 0.0,
+    ) -> bool:
+        """Save the learner state.
+
+        Args:
+            timestep (int):
+                timestep at which the state is being saved.
+            unreplicated_learner_state (Union[LearnerState, RNNLearnerState]):
+                a Mava LearnerState (must be unreplicated)
+            episode_return (Numeric, optional):
+                Optional value to determine whether this is the 'best' model to save.
+                Defaults to 0.0.
+
+        Returns:
+            bool: whether the saving was successful.
+        """
+        model_save_success: bool = self._manager.save(
+            step=timestep,
+            items={
+                "learner_state": unreplicated_learner_state,
+                "type": re.findall(
+                    "\<class '(.*)'\>",  # noqa: W605
+                    str(type(unreplicated_learner_state)),
+                )[0],
+            },
+            # TODO: Currently we only log the episode return,
+            #       but perhaps we should log other metrics.
+            metrics={"episode_return": episode_return},
+        )
+        return model_save_success
+
+    def restore_learner_state(
+        self,
+        unreplicated_input_learner_state: Union[LearnerState, RNNLearnerState],
+        n: Optional[int] = None,
+        restore_params: bool = True,
+        restore_hstates: bool = True,
+    ) -> Union[LearnerState, RNNLearnerState]:
+        """Restore the learner state.
+
+        Args:
+            n (Optional[int], optional):
```

Comment:
I'm finding it difficult to find a nice name for this argument. But many something like `restart_env_timestep`?

Reply:
Changed to `timestep` as it refers to the same thing as `timestep` in the `save()` method.